cmd/vicadmin/server.go

```diff
@@ -112,39 +127,208 @@ func (s *server) listenPort() int {
+		// take the form data and use it to try to authenticate with vsphere
+
+		// create a userconfig
+		userconfig := session.Config{
+			Insecure:       false,
+			Thumbprint:     rootConfig.Thumbprint,
+			Keepalive:      rootConfig.Keepalive,
+			ClusterPath:    rootConfig.ClusterPath,
+			DatacenterPath: rootConfig.DatacenterPath,
+			DatastorePath:  rootConfig.DatastorePath,
+			HostPath:       rootConfig.Config.HostPath,
+			PoolPath:       rootConfig.PoolPath,
+		}
+		user := url.UserPassword(req.FormValue("username"), req.FormValue("password"))
+		serviceURL, err := soap.ParseURL(rootConfig.Service)
+		if err != nil {
+			// this could happen for a number of reasons but most likely for a plain ol' auth failure
+			log.Errorf("vSphere service URL was not a valid format; parsing returned error: %s", err)
+			http.Error(res, genericErrorMessage, http.StatusInternalServerError)
+			return
+		}
+
+		serviceURL.User = user
+		userconfig.Service = serviceURL.String()
+
+		// check login
+		usersession, err := vSphereSessionGet(&userconfig)
+		if err != nil || usersession == nil {
+			// something went wrong or we could not authenticate
+			log.Warnf("User %s from %s failed to authenticated at %s", user, req.RemoteAddr, time.Now())
+			http.Error(res, "Authentication failed due to incorrect credential(s)", 400)
+			return
+		}
+
+		// successful login above; user is authenticated
+		// log out, disregard errors
+		usersession.Client.Logout(context.Background())
+
+		// save user config locally
+		usersess := s.uss.Add(req.FormValue("username"), &userconfig)
+
+		// create a token to save as an encrypted & signed cookie
+		websession, err := s.uss.cookies.Get(req, sessionCookieKey)
+		if websession == nil {
+			log.Errorf("Web session object could not be created due to error %s", err)
+			http.Error(res, genericErrorMessage, http.StatusInternalServerError)
+			return
+		}
+
+		timeNow, err := usersess.created.MarshalText()
+		if err != nil {
+			log.Errorf("Failed to unmarshal time object %+v into text due to error: %s", usersess.created, err)
+			http.Error(res, genericErrorMessage, http.StatusInternalServerError)
+			return
+		}
+
+		websession.Values[sessionCreationTimeKey] = string(timeNow)
+		websession.Values[sessionUsernameKey] = req.FormValue("username")
+		if err := websession.Save(req, res); err != nil {
+			log.Errorf("\"%s\" occurred while trying to save session to browser", err.Error())
+			http.Error(res, genericErrorMessage, http.StatusInternalServerError)
+			return
+		}
+
+		// redirect to dashboard
+		http.Redirect(res, req, "/", http.StatusTemporaryRedirect)
+		return
+	}
+
+	// Render login page (shows up on non-POST requests):
+	sess, err := client(&rootConfig)
+	if err != nil {
+		log.Errorf("Could not render login page due to vSphere connection error: %s", err.Error())
+		http.Error(res, genericErrorMessage, http.StatusInternalServerError)
+		return
+	}
+	v := vicadmin.NewValidator(ctx, &vchConfig, sess)
```

Comment:
We should get this done for 0.8 without doubt. If it doesn't go in this PR, please open an issue to do so as it could easily slip through cracks otherwise.